### PR TITLE
feat: WebAuthn passkey authentication

### LINKED ISSUE #19: Feature: PassKey Authentication
# Implement WebAuthn/Passkey Authentication for PlayaPlan

## Overview

Add support for WebAuthn/passkey authentication to improve security and user experience. Passkeys will provide a more secure, phishing-resistant authentication method alongside our existing email verification code system.

## Goals

- Implement WebAuthn registration for existing users
- Support passwordless login with WebAuthn passkeys
- Maintain backward compatibility with email code verification
- Allow users to manage their passkeys
- Ensure cross-browser compatibility where WebAuthn is supported

### System Components

```mermaid
graph TD
    User[User Device] -->|Registers/Authenticates| Frontend[Frontend React App]
    Frontend -->|API Calls| Backend[NestJS Backend]
    Backend -->|Query/Store| Database[(PostgreSQL Database)]
    
    subgraph "Frontend Components"
        PasskeyLoginComponent[Passkey Login Component]
        PasskeyRegistrationComponent[Passkey Registration Component]
        PasskeyManagementComponent[Passkey Management Component]
        AuthContext[Auth Context]
    end
    
    subgraph "Backend Components"
        WebAuthnController[WebAuthn Controller]
        WebAuthnService[WebAuthn Service]
        WebAuthnStrategy[WebAuthn Passport Strategy]
        AuthService[Auth Service]
    end
```

### Authentication Flow

```mermaid
sequenceDiagram
    participant User
    participant Frontend
    participant Backend
    participant Database
    
    %% Registration
    User->>Frontend: Initiates Passkey Registration
    Frontend->>Backend: Request Registration Options
    Backend->>Database: Create Registration Challenge
    Backend->>Frontend: Return Registration Options
    Frontend->>User: Prompt for Biometric/PIN
    User->>Frontend: Provides Biometric/PIN
    Frontend->>Backend: Send Registration Response
    Backend->>Database: Verify and Store Credential
    Backend->>Frontend: Registration Success
    
    %% Authentication
    User->>Frontend: Initiates Passkey Login
    Frontend->>Backend: Request Authentication Options
    Backend->>Database: Create Authentication Challenge
    Backend->>Frontend: Return Authentication Options
    Frontend->>User: Prompt for Passkey
    User->>Frontend: Provides Passkey
    Frontend->>Backend: Send Authentication Response
    Backend->>Database: Verify Credential
    Backend->>Frontend: Return JWT Token
    Frontend->>Frontend: Set Authentication State
```

### Database Changes

Add the following to the Prisma schema:

```prisma
model UserCredential {
  id              String   @id @default(uuid())
  userId          String
  user            User     @relation(fields: [userId], references: [id])
  credentialId    String   @unique // Base64URL-encoded credential ID
  publicKey       Bytes    // Credential public key in COSE format
  counter         Int      // Signature counter for detecting cloned authenticators
  transports      String?  // JSON array of transports (platform, cross-platform)
  algorithm       String?  // Signature algorithm identifier
  attestationFormat String? // Format of attestation (none, indirect, etc.)
  createdAt       DateTime @default(now())
  updatedAt       DateTime @updatedAt

  @@index([userId])
}

// Update User model
model User {
  // existing fields...
  
  credentials     UserCredential[]
  hasPasskeys     Boolean  @default(false)
}
```

### Backend Requirements

1. **Dependencies**
   - Add `@simplewebauthn/server` package
   - Add `base64url` package

2. **New API Endpoints**
   - `/auth/webauthn/register/options` - Generate registration options
   - `/auth/webauthn/register/verify` - Verify registration response
   - `/auth/webauthn/login/options` - Generate authentication options
   - `/auth/webauthn/login/verify` - Verify authentication response
   - `/auth/webauthn/credentials` - List user credentials
   - `/auth/webauthn/credentials/:id` - Delete credential

3. **Services**
   - Create `WebAuthnService` to handle WebAuthn operations
   - Update `AuthService` to integrate with WebAuthn flow

4. **Configuration**
   - Add WebAuthn-specific configuration (RP name, ID, origins)

### Frontend Requirements

1. **Dependencies**
   - Add `@simplewebauthn/browser` package

2. **Auth Context Updates**
   - Add passkey detection/support
   - Add methods for registering and authenticating with passkeys
   - Track user passkey status

3. **New Components**
   - Create passkey login component
   - Create passkey registration component
   - Create passkey management UI in user profile

4. **Auth Flow Updates**
   - Update login flow to offer passkey option when available
   - Show appropriate UI for browsers with/without WebAuthn support

## Security Considerations

- Properly validate credential counter values to prevent replay attacks
- Use unique challenges for each operation
- Maintain email verification as fallback authentication method
- Implement clear user flows for account recovery
- Add comprehensive logging for security auditing

## Testing Approach

- Unit tests for WebAuthn service and controller
- Integration tests for authentication flows
- Browser compatibility testing
- Security testing for edge cases

## Acceptance Criteria

- [ ] Users can register passkeys to their accounts
- [ ] Users can authenticate using passkeys
- [ ] Users can manage (view/delete) their passkeys
- [ ] Browsers without WebAuthn support gracefully fall back to email verification
- [ ] All WebAuthn operations are properly secured against known attacks
- [ ] Documentation is updated with passkey usage instructions
- [ ] Test coverage is maintained for new code

### Backend Implementation

```typescript
// apps/api/src/auth/services/webauthn.service.ts

@Injectable()
export class WebAuthnService {
  private rpName: string;
  private rpID: string;
  private origin: string;
  
  constructor(
    private readonly prisma: PrismaService,
    private readonly configService: ConfigService,
    private readonly jwtService: JwtService,
  ) {
    // Initialize from config
    this.rpName = this.configService.get<string>('webauthn.rpName', 'PlayaPlan');
    this.rpID = this.configService.get<string>('webauthn.rpID', 'localhost');
    this.origin = this.configService.get<string>('webauthn.origin', 'http://localhost:3000');
  }
  
  /**
   * Generate registration options for a user
   */
  async generateRegistrationOptions(userId: string): Promise<GenerateRegistrationOptionsOpts> {
    // Implementation...
  }
  
  /**
   * Verify registration response
   */
  async verifyRegistration(userId: string, credential: RegistrationCredentialJSON): Promise<boolean> {
    // Implementation...
  }
  
  /**
   * Generate authentication options for a user
   */
  async generateAuthenticationOptions(email: string): Promise<GenerateAuthenticationOptionsOpts> {
    // Implementation...
  }
  
  /**
   * Verify authentication response
   */
  async verifyAuthentication(
    credential: AuthenticationCredentialJSON,
    expectedChallenge: string
  ): Promise<User> {
    // Implementation...
  }
}
```

### Frontend Implementation

```typescript
// In AuthContext.tsx
const registerPasskey = async (): Promise<boolean> => {
  setError(null);
  setIsLoading(true);
  
  try {
    if (!user?.email) {
      throw new Error('User must be logged in to register a passkey');
    }
    
    // Get registration options from server
    const options = await webauthn.getRegistrationOptions(user.email);
    
    // Request credential creation from browser
    const credential = await startRegistration(options);
    
    // Verify registration with server
    await webauthn.verifyRegistration(user.email, credential);
    
    // Update user state to reflect passkeys are available
    setHasPasskeys(true);
    
    return true;
  } catch (err) {
    const errorMessage = err instanceof Error 
      ? err.message 
      : 'Failed to register passkey';
    
    setError(errorMessage);
    console.error('Passkey registration failed:', err);
    return false;
  } finally {
    setIsLoading(false);
  }
};
```

## Resources

- [WebAuthn Guide](https://webauthn.guide/)
- [SimpleWebAuthn Documentation](https://simplewebauthn.dev/)
- [FIDO Alliance WebAuthn Overview](https://fidoalliance.org/passkeys/)

### PR DESCRIPTION
**Pull request by ChrisRomp**
## Summary
WebAuthn passkey authentication for PlayaPlan: usernameless ("Sign in with a passkey") login plus per-user passkey enrollment and management from the profile page.

### What it does
- New API module `apps/api/src/passkeys` with registration + assertion ceremonies built on `@simplewebauthn/server`
- Two new Prisma models: `Passkey` (per credential) and `WebAuthnChallenge` (server-stored, single-use, TTL-bounded)
- Boot-time `WEBAUTHN_*` config validator (RP ID consistency with origin, HTTPS required outside loopback)
- Frontend: `usePasskeys` hook, `PasskeysSection` profile UI, "Sign in with a passkey" button on `LoginForm`, `AuthContext` plumbing
- Email-verification gate on enrollment so a passkey cannot be added before the user proves email ownership
- Rate limiting on auth endpoints (per-throttler `skipIf` routing in `@nestjs/throttler@6`)
- RFC 8176 `amr` claim on JWT: `["otp"]` for email-code login, `["hwk", "mfa"]` for passkey assertion (hardware-key + multi-factor; the UV requirement satisfies MFA)

### Key changes
- `apps/api/prisma/schema.prisma` + new migration: `Passkey`, `WebAuthnChallenge` models
- `apps/api/src/passkeys/services/passkeys.service.ts`: full WebAuthn ceremony  challenge create/consume, monotonic counter persistence with concurrent-write race re-read, Serializable-tx capacity check, P2002 normalization, malformed-payload normalizationlogic 
- `apps/api/src/passkeys/controllers/passkey.controller.ts`: REST endpoints with `assertVerified()` gate
- `apps/api/src/auth/controllers/auth.controller.ts`: passkey options/verify endpoints + `amr` claim plumbing
- `apps/api/src/common/throttling/`: rewritten for `@nestjs/throttler@6` API (`shouldSkip` / `getTracker` / `generateKey`); per-throttler `skipIf` routes auth requests to the strict bucket and everything else to the relaxed bucket; APP_GUARD now consumes the same options object that `forRootAsync` produces (`getOptionsToken()` injection) so `THROTTLE_TTL` / `THROTTLE_LIMIT` env vars actually take effect
- `apps/api/src/main.ts`: `app.set('trust proxy', 1)` so `req.ip` reflects forwarded client IP
- `apps/api/src/config/webauthn-config.validator.ts`: boot-time validator. Loopback exception covers both `localhost` and `127.0.0.1` for both origin protocol (HTTP allowed) and rpId (IP-as-rpId allowed when origin matches)
- `apps/web/src/hooks/usePasskeys.ts` + `apps/web/src/lib/api/passkeys.ts`: React hook and API client
- `apps/web/src/components/profile/PasskeysSection.tsx`: per-user list + add/rename/delete UI
- `apps/web/src/components/auth/LoginForm.tsx`: passkey login button gated on basic WebAuthn support (`window.PublicKeyCredential` present). The button drives an explicit modal flow rather than the autofill conditional-mediation flow; this avoids depending on `isConditionalMediationAvailable()` while still matching the user's expectations across browsers
- `apps/web/src/store/AuthContext.tsx`: passkey login wired through existing token/session flow
- `apps/web/src/lib/api.ts`: 401 interceptor skips `/auth/passkey/options` and `/auth/passkey/verify`

### Security
- Server-stored single-use challenges (TTL 5  never trust browser-supplied challengemin) 
- `requireUserVerification: true` on both registration and assertion
- Conditional `updateMany(counter < new)` for monotonic counter; race-loser re-reads and rejects on regression (clone detection)
- Serializable transaction around capacity check + create
- Per-user `MAX_PASSKEYS_PER_USER` cap
- Email-verification gate on enrollment closes the legacy `/auth/register` pre-verification window
- Boot rejects non-`https://` `WEBAUTHN_ORIGIN` outside `localhost` / `127.0.0.1`
- 12-char SHA-256 hash of credential IDs in unknown-credential warn logs (no raw credential identifiers in logs)
- Throttler buckets driven by env (`THROTTLE_TTL` / `THROTTLE_LIMIT`) plus `trust proxy` so IP-based limits work behind a reverse proxy
- E2E round-trip test with a mocked WebAuthn verifier in `apps/api/test/passkeys.e2e-spec.ts`

### Tests
- 843 API tests pass
- 601 web tests pass
- Lint clean, build clean
- Manual smoke test: enroll + login round-trip works through SSH-tunneled dev environment

Fixes [#19](https://github.com/ChrisRomp/playa-plan/issues/19)

### CODE REVIEW

File: apps/api/src/passkeys/services/passkeys.service.ts [hunk 1]
```diff
@@ -0,0 +1,627 @@
...
+      rpName: cfg.rpName,
+      rpID: cfg.rpId,
+      userName: user.email,
+      userID: userIdBytes,
+      userDisplayName: this.displayName(user),
+      attestationType: 'none',
+      excludeCredentials: existing.map((c) => ({
+        id: c.credentialId,
+        transports: c.transports as never,
+      })),
```
[the remark this answers is not in the document]
Comment (ChrisRomp) [reply apps/api/src/passkeys/services/passkeys.service.ts:103]: The v9 API took raw `Buffer`/`Uint8Array` IDs and required `type: 'public-key'`, but in `@simplewebauthn/server` v13 (what we're on) `excludeCredentials[].id` is typed as `Base64URLString` and `type` is no longer part of the descriptor — see `node_modules/@simplewebauthn/server/esm/registration/generateRegistrationOptions.d.ts`. The library generates the JSON-form options and the `PublicKeyCredentialDescriptorJSON` shape it ships back already includes `type: 'public-key'` for the browser. So passing the stored base64url `credentialId` directly is correct and exclusion works as expected.

The `as never` cast was sloppy though; I replaced it with `as AuthenticatorTransportFuture[]` so the transport array is properly typed instead of being widened away.

File: apps/api/src/main.ts [hunk 2]
```diff
@@ -1,15 +1,29 @@
...
+  const app = await NestFactory.create<NestExpressApplication>(AppModule);
   const configService = app.get(ConfigService);
-  
+
+  // Trust the first reverse proxy in front of the API so that `req.ip` and
+  // helmet's secure-context detection use the X-Forwarded-* values rather
+  // than the proxy's loopback address. Without this, throttling buckets all
+  // real clients behind a single proxy IP and CSP / cookie security
+  // heuristics may misbehave under HTTPS termination at the proxy.
+  app.set('trust proxy', 1);
```
[the remark this answers is not in the document]
Comment (ChrisRomp) [reply apps/api/src/main.ts]: Good call. Made `app.set('trust proxy', ...)` env-driven via a new `TRUST_PROXY` var (parsed by `parseTrustProxy` in `main.ts`), and flipped the default to `false` so a direct-exposure deployment can't be tricked by spoofed `X-Forwarded-For`. Production deployments behind a reverse proxy now have to opt in with `TRUST_PROXY=1` (or higher for nested proxies). Documented in `.env.sample` and added to the Joi validation schema, plus unit tests for the parser.

File: apps/api/src/passkeys/services/passkeys.service.ts [hunk 3]
```diff
@@ -0,0 +1,628 @@
...
+    let row;
+    try {
+      // Atomic single-use: delete by unique field. If two requests race,
+      // exactly one succeeds; the other gets P2025 and we surface a clean
+      // BadRequestException instead of a 500.
+      row = await this.prisma.webAuthnChallenge.delete({
+        where: { challenge },
+      });
+    } catch {
+      throw new BadRequestException('Unknown or already-used challenge');
```
[the remark this answers is not in the document]
Comment (ChrisRomp) [reply apps/api/src/passkeys/services/passkeys.service.ts]: Good catch. Narrowed the catch to `Prisma.PrismaClientKnownRequestError` with code `P2025`; everything else (connection loss, permission failures, etc.) now rethrows so it surfaces as a 5xx and is visible in monitoring. Added a unit test that confirms a `P1001` propagates through `verifyRegistration`, and updated the existing P2025 test to throw a real `PrismaClientKnownRequestError`.

File: apps/api/prisma/migrations/20260428232541_add_passkeys/migration.sql [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+-- CreateTable
+CREATE TABLE "passkeys" (
+    "id" TEXT NOT NULL,
+    "userId" TEXT NOT NULL,
+    "credentialId" TEXT NOT NULL,
+    "publicKey" BYTEA NOT NULL,
+    "webAuthnUserID" TEXT NOT NULL,
+    "counter" BIGINT NOT NULL DEFAULT 0,
+    "transports" TEXT[],
```
[the remark this answers is not in the document]
Comment (ChrisRomp) [reply apps/api/prisma/migrations/20260428232541_add_passkeys/migration.sql]: Good catch — the migration is only on this feature branch (not yet on main), so I edited it in place: `transports` is now `TEXT[] NOT NULL DEFAULT ARRAY[]::TEXT[]`. Also added `@default([])` to the Prisma model so `prisma migrate diff` stays clean and code paths that create a passkey without explicitly setting transports get an empty array instead of a null/missing value. Verified with a fresh `prisma migrate reset` + full test suite (853 passing).
